lib/src/util/lint_cache.dart

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2019, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:analyzer/dart/analysis/analysis_context_collection.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/file_system/physical_file_system.dart';
+
+import '../analyzer.dart';
+import '../rules.dart';
+import '../utils.dart';
+
+const String _INCOMPATIBLE_WITH_CLASS_NAME = 'IncompatibleWith';
+const String _LINT_RULE_CLASS_NAME = 'LintRule';
+const String _LINTER_META_LIB_NAME = 'linter_meta';
+
+bool _isIncompatibleWithAnnotation(Element element) =>
+    element is ConstructorElement &&
+    element.enclosingElement.name == _INCOMPATIBLE_WITH_CLASS_NAME &&
+    element.library?.name == _LINTER_META_LIB_NAME;
+
+class LintCache {
+  List<LintRuleDetails> details = <LintRuleDetails>[];
+
+  LintRuleDetails findDetailsByClassName(String className) {
+    for (var detail in details) {
+      if (detail.className == className) {
+        return detail;
+      }
+    }
+    return null;
+  }
+
+  LintRuleDetails findDetailsById(String id) {
+    for (var detail in details) {
+      if (detail.id == id) {
+        return detail;
+      }
+    }
+    return null;
+  }
+
+  Future<void> init() async {
+    registerLintRules();
+
+    // Setup details.
+    for (var lint in Analyzer.facade.registeredRules) {
+      details.add(LintRuleDetails(lint.runtimeType.toString(), lint.name));
+    }
+
+    // Process compatibility annotations.
+    final rulePath = File('lib/src/rules').absolute.path;
```

Comment:
Alternatively, we could analyze and visit the constructor invocations in `lib/src/rules.dart`. That would probably find all of the lint rule class element faster (although it would miss classes that aren't registered).

Reply:
Let's discuss refining this together.  I have a few other ideas but a whiteboard would make them easier.

I'll land this as a jumping off point.